service/sharddistributor/handler/handler.go

```diff
@@ -0,0 +1,76 @@
+// The MIT License (MIT)
+
+// Copyright (c) 2017-2020 Uber Technologies Inc.
+
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package handler
+
+import (
+	"context"
+	"sync"
+
+	"github.com/uber/cadence/client/matching"
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/metrics"
+	"github.com/uber/cadence/common/types"
+)
+
+func NewHandler(
+	logger log.Logger,
+	metricsClient metrics.Client,
+	matchingPeerResolver matching.PeerResolver,
+) Handler {
+	handler := &handlerImpl{
+		logger:               logger,
+		metricsClient:        metricsClient,
+		matchingPeerResolver: matchingPeerResolver,
+	}
+
+	// prevent us from trying to serve requests before matching engine is started and ready
```

Comment:
what is the matching engine in this context?

Reply:
@jakobht  what do you think about adding an arch diagram sketch into the `docs/arch` folder to contextualize what you're thinking / be able to point those of us (such as myself) for whom context may have fallen out of our brains.

Adding the mapper/handler shell all seems all good, but it'd probably help with review

Reply:
@Shaddoll Sorry bad copy paste fixed the comment. 

@davidporter-id-au I think thats a good idea, added a task to do that